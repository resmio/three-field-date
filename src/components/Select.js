import React from 'react'
import PropTypes from 'prop-types'
import { css } from 'glamor'
import { colors } from '@resmio/rollico/dist'

const field = (hasError) => css({
  background: colors.white,
  fontSize: '1em',
  height: '2.4em',
  color: colors.dustyGray,
  borderColor: hasError ? colors.amaranth : colors.dustyGray,
  ':focus': {
    outline: 'none'
  }
})

const Select = ({
  id,
  name,
  onOptionChange,
  options,
  selected,
  values,
  hasError
}) => (
  <select {...field(hasError)} name={name} value={selected} onChange={onOptionChange} id={id} >
    <option value={0}>{name}</option>
    {
      options.map((option, i) => (
        <option key={i} value={values ? values[i] : option}>{option}</option>
      ))
    }
  </select>
)

const {
  arrayOf,
  bool,
  func,
  number,
  oneOfType,
  string
} = PropTypes

Select.propTypes = {
  hasError: bool,
  id: string,
  name: string,
  onOptionChange: func.isRequired,
  options: arrayOf(oneOfType([string, number])).isRequired,
  selected: oneOfType([string,number]),
  values: arrayOf(oneOfType([string, number]))
}

export default Select
