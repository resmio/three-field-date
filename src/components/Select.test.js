import React from 'react'
import renderer from 'react-test-renderer'

import Select from './Select'

const ID = 'testId'
const NAME = 'testName'
const OPTIONS = ['1','2','3','4']
const VALUES = ['value1', 'value2', 'value3', 'value4']
const SELECTED = '3'
const ON_OPTION_CHANGE = jest.fn()

describe('Select', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Select
          id={ID}
          name={NAME}
          onOptionChange={ON_OPTION_CHANGE}
          options={OPTIONS}
          selected={SELECTED}
        />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders values if those are passed as props', () => {
    const tree = renderer
      .create(
        <Select
          onOptionChange={ON_OPTION_CHANGE}
          options={OPTIONS}
          selected={SELECTED}
          values={VALUES}
        />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
  it('renders differently if it has an error', () => {
    const tree = renderer
      .create(
        <Select
          onOptionChange={ON_OPTION_CHANGE}
          options={OPTIONS}
          selected={SELECTED}
          hasError
        />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should fire the onOptionChange prop passing the value selected', () => {
    const tree = renderer
      .create(
        <Select
          onOptionChange={ON_OPTION_CHANGE}
          options={OPTIONS}
          selected={SELECTED}
        />).toJSON()

    tree.props.onChange()

    expect(tree).toMatchSnapshot()
    expect(ON_OPTION_CHANGE).toBeCalled()
  })
})
