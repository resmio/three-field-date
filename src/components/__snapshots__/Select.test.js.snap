exports[`Select renders correctly 1`] = `
<select
  data-css-8zqqrv=""
  id="testId"
  name="testName"
  onChange={[Function]}
  value="3">
  <option
    value={0}>
    testName
  </option>
  <option
    value="1">
    1
  </option>
  <option
    value="2">
    2
  </option>
  <option
    value="3">
    3
  </option>
  <option
    value="4">
    4
  </option>
</select>
`;

exports[`Select renders differently if it has an error 1`] = `
<select
  data-css-pmxpya=""
  id={undefined}
  name={undefined}
  onChange={[Function]}
  value="3">
  <option
    value={0} />
  <option
    value="1">
    1
  </option>
  <option
    value="2">
    2
  </option>
  <option
    value="3">
    3
  </option>
  <option
    value="4">
    4
  </option>
</select>
`;

exports[`Select renders values if those are passed as props 1`] = `
<select
  data-css-8zqqrv=""
  id={undefined}
  name={undefined}
  onChange={[Function]}
  value="3">
  <option
    value={0} />
  <option
    value="value1">
    1
  </option>
  <option
    value="value2">
    2
  </option>
  <option
    value="value3">
    3
  </option>
  <option
    value="value4">
    4
  </option>
</select>
`;

exports[`Select should fire the onOptionChange prop passing the value selected 1`] = `
<select
  data-css-8zqqrv=""
  id={undefined}
  name={undefined}
  onChange={[Function]}
  value="3">
  <option
    value={0} />
  <option
    value="1">
    1
  </option>
  <option
    value="2">
    2
  </option>
  <option
    value="3">
    3
  </option>
  <option
    value="4">
    4
  </option>
</select>
`;
